update_cpes.py

```diff
@@ -7,45 +7,117 @@
 import yaml
 from lxml import etree
 
+# CPE w/o 2.3 component: cpe:/a:nginx:nginx:0.1.0"
+REGEX_CPE = re.compile('^cpe:/([aho]):([^:]+):([^:]+)')
+# CPE w/  2.3 component: cpe:2.3:a:f5:nginx:0.1.0:*:*:*:*:*:*:*
+REGEX_CPE_23 = re.compile('^cpe:2.3:([aho]):([^:]+):([^:]+)')
+
+XML_PATH_DEPRECATED_BY = "./{http://scap.nist.gov/schema/cpe-extension/2.3}cpe23-item/{http://scap.nist.gov/schema/cpe-extension/2.3}deprecation/{http://scap.nist.gov/schema/cpe-extension/2.3}deprecated-by"
+
 def parse_r7_remapping(file):
     with open(file) as remap_file:
         return yaml.safe_load(remap_file)["mappings"]
 
+def update_vp_map(target_map, cpe_type, vendor, product):
+    """Add an entry to the dict tracking valid combinations
+    """
+
+    if cpe_type not in target_map:
+        target_map[cpe_type] = {}
+
+    if vendor not in target_map[cpe_type]:
+        target_map[cpe_type][vendor] = set()
+
+    product = product.replace('%2f', '/')
+    target_map[cpe_type][vendor].add(product)
+
+
+def update_deprecated_map(target_map, dep_string, entry):
+    """Add an entry to the dict tracking deprecations
+
+    target_map example:
+
+    {
+      "a:100plus:101eip":
+        {
+          "deprecated_date": "2021-06-10T15:28:05.490Z",
+          "deprecated_by": "a:hundredplus:101eip"
+        }
+    }
+
+    Args:
+        target_map (dict): dict containing deprecations
+        dep_string (str): key to add in the format of 'type:vendor:product'
+        entry (lxml.etree._Element): XML element to pull additional data from
+
+    Returns:
+        Nothing, target_map modified in place
+    """
+
+    deprecated_date = entry.get("deprecation_date", "")
+
+    # Find the CPE that deprecated this entry
+    raw_dep_by = entry.find(XML_PATH_DEPRECATED_BY).get('name')
+
+    # Extract the type, vendor, product
+    dep_by_match = REGEX_CPE_23.match(raw_dep_by)
+    if not dep_by_match:
+        logging.error("CPE %s is deprecated but we can't build the deprecation mapping entry for some reason.", dep_string)
+        return
+
+    dep_type, dep_vendor, dep_product = dep_by_match.group(1, 2, 3)
+    deprecated_by = "{}:{}:{}".format(dep_type, dep_vendor, dep_product)
+
+    if dep_string not in target_map:
+        target_map[dep_string] = {}
+
+    if not target_map[dep_string].get('deprecated_date'):
```

Comment:
I've seen different deprecated date values for type:vendor:product values. In this case first one processed wins. Is there any value in storing the `deprecated_date` and `deprecated_by` tuples?